### DISCUSSION TITLE
Pylint: E1133:not-an-iterable on query result

### DISCUSSION THREAD

**Discussion by davetapley** [Textualize/textual#3671] [General]
**Edit**:

Caused by upstream bug:
- https://github.com/pylint-dev/astroid/issues/1015

Workaround is to use [`results`](https://textual.textualize.io/api/query/#textual.css.query.DOMQuery.results).

---

Can anyone replicate this?

```py
inputs = self.query(Input)
ids = [input.id for input in inputs] # Pylint: E1133:not-an-iterable 
```

Makes no sense because:
https://github.com/Textualize/textual/blob/81531e87647eae26b83ad31f1941016b4f2bab41/src/textual/css/query.py#L141-L142

And following though Pylint code it seems that's all it should need:

1. https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2291-L2294
1. https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2250-L2257
1. https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/utils.py#L1298-L1302
1. https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/utils.py#L1223

Any ideas? 🤔

**Comment by davetapley** [comment]
Debugging I figured out it gets down to [here](https://github.com/pylint-dev/astroid/blob/2d309cbb823e27d73bf695447a50f7875a852612/astroid/nodes/node_ng.py#L126-L181) and gets `Const.NoneType` for `inputs`. I don't know why, but I _assume_ it then yields error because `None` isn't iterable.

... except none of `Query` return `None` 🤯 
https://github.com/Textualize/textual/blob/81531e87647eae26b83ad31f1941016b4f2bab41/src/textual/dom.py#L1073-L1083

**Comment by davep** [comment]
> Can anyone replicate this?

Always. It's been one of those things where I see it while I'm in the middle of something else, go "I really should make a note of that but I'm busy right now", and then forget once I'm done with the code I'm writing. It's been bugging me for ages now and I keep meaning to check what's what.

The short (but unsatisfying) answer is "use [`results`](https://textual.textualize.io/api/query/#textual.css.query.DOMQuery.results)"; would be good to get to the bottom of that though.

**Comment by davetapley** [comment]
@davep well I'm glad it's not just me 😁 
Thanks for workaround.

I did some digging prior to seeing your reply, my notes if someone else wants to pick it up:

I use VSCode debug with `launch.json` (replace `file.py` with a file with a `query` in it):
```
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "pylint",
            "type": "python",
            "request": "launch",
            "module": "pylint",
            "justMyCode": false,
            "args": ["file.py"]
        },
```

Then break point here: https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2291-L2294

Add disabled break point here:
https://github.com/pylint-dev/astroid/blob/2d309cbb823e27d73bf695447a50f7875a852612/astroid/decorators.py#L94

Hit first break, then enable second break and continue.

Will hit `raise InferenceError(**error.args[0])` with something like:

```
error.args[0]
{'node': <Comprehension l.73 ...7bc9233d0>, 'unknown': <AssignName.i l.73 a...7bc923450>, 'assign_path': None, 'context': <astroid.context.Inf...7ade519c0>}
```

which causes `yield Uninferable`, then `inferred = Const.NoneType`,
and [here](https://github.com/pylint-dev/pylint/blob/a841a7d9eab654e423a115b4feaa857d5f807c88/pylint/checkers/typecheck.py#L2253-L2257) gives `not is_iterable(inferred,..` 😞 


But that's so deep in to `astroid` internals I have no idea why `InferenceError` is raised 🤔

**Reply by davetapley** [reply]
Posted to https://discuss.python.org/t/reproducible-but-incorrect-pylint-e1133-not-an-iterable-in-library/38633, and:
code-quality@python.org

So 🤞🏻

**Comment by davetapley** [comment]
@davep we have an answer, per:
- https://github.com/pylint-dev/pylint/issues/9253#issuecomment-1826423574

Underlying issue is:
- https://github.com/pylint-dev/astroid/issues/1015

But currently unfixed and no workaround 😞
